src/io/proxi.rs

```diff
@@ -407,29 +618,121 @@ where
     deserializer.deserialize_any(USIVisit {})
 }
 
+use serde::de::{self, Visitor};
+
+/// MassIVE returns a list of strings instead of a list of numbers, this type can be deserialized if a number of string is given in the JSON
+#[derive(Debug, Default, Clone)]
+pub struct Wrapped<T>(T);
+
+impl<T> std::ops::Deref for Wrapped<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'de, T: 'static + Default + Copy + FromStr> serde::Deserialize<'de> for Wrapped<T>
+where
+    T::Err: Display,
+    f64: AsPrimitive<T>,
+    u64: AsPrimitive<T>,
+    i64: AsPrimitive<T>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        deserializer
+            .deserialize_any(PotentiallyWrappedNumberVisitor::<T>::default())
+            .map(Wrapped)
+    }
+}
+
+impl serde::Serialize for Wrapped<f64> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_f64(self.0)
+    }
+}
+
+impl serde::Serialize for Wrapped<f32> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_f32(self.0)
+    }
+}
+
+#[derive(Debug, Default)]
+struct PotentiallyWrappedNumberVisitor<T> {
+    marker: PhantomData<T>,
+}
+
+impl<'de, T: 'static + Copy + FromStr> Visitor<'de> for PotentiallyWrappedNumberVisitor<T>
+where
+    T::Err: Display,
+    f64: AsPrimitive<T>,
+    u64: AsPrimitive<T>,
+    i64: AsPrimitive<T>,
+{
+    type Value = T;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a wrapped number")
+    }
+
+    fn visit_f64<E: de::Error>(self, value: f64) -> Result<Self::Value, E> {
+        Ok(value.as_())
+    }
+
+    fn visit_i64<E: de::Error>(self, value: i64) -> Result<Self::Value, E> {
+        Ok(value.as_())
+    }
+
+    fn visit_u64<E: de::Error>(self, value: u64) -> Result<Self::Value, E> {
+        Ok(value.as_())
+    }
+
+    fn visit_str<E: de::Error>(self, value: &str) -> Result<Self::Value, E> {
+        value.parse().map_err(|e| serde::de::Error::custom(e))
+    }
+}
+
+/// A spectrum returnd by a PROXI server
 #[derive(Debug, Default, Clone, Serialize, Deserialize)]
 pub struct PROXISpectrum {
-    #[serde(serialize_with = "usi_serialize", deserialize_with = "usi_deserialize")]
+    /// The USI as returned by the PROXI server
+    #[serde(
+        default,
+        serialize_with = "usi_serialize",
+        deserialize_with = "usi_deserialize"
+    )]
     pub usi: Option<USI>,
+    /// The status of this request
     pub status: Option<Status>,
+    /// Metadata for this spectrum
+    #[serde(default)]
     pub attributes: Vec<PROXIParam>,
     #[serde(default)]
-    pub mzs: Vec<f64>,
+    pub mzs: Vec<Wrapped<f64>>,
```

Comment:
I have fought with serde a bit more (already tried this for the first implementation) and I do not think I can finagle serde to accept what I want. The problem here is the `Vec`, I cannot just say deserialize_with my wrapped deserializer because that would result in `f64` instead of `Vec<f64>`. Building the stuff to handle the full vec derserialization turned out to be a lot more complicated than I hoped so I gave up. 

Is it needed to have a `Vec<f64>` for performance, or could a `Iterator<Item=f64>` work? The last one is easy with safe rust (`mzs.iter().copied()`), and the first one could be made but might need some `mem::transmute` or similar. I was thinking of removing the pub specifier for the struct field and creating a function that returns the data in a way that is more convenient for use within mzdata.

But of course if you have experience with serde having serde return the right thing would be best.

Reply:
This requires a two visitors, one to implement the number-or-string deserialization and one to implement `deserialize_seq` using the first visitor. I can write that implementation later today or tomorrow, time permitting.

Reply:
Done. `deserialize_seq` is quite powerful, letting you do all kinds of things like filter data s you deserialize it.

Accessors would probably not be a bad idea. I don't use them until they are needed for abstraction, which for API design might qualify as "too late". On the other hand, `PROXISpectrum` is meant to be a useable intermediary state, not a full `SpectrumLike` implementation.

Reply:
Thanks! It was nice to see how you did it, I tried something similar yesterday but was bogged down by the `SeqAccess` because I thought I needed to do something with that. This is the first time I actually use Serde for something complex so it was a nice learning opportunity.